### PR TITLE
Empire daemon

### PR DESCRIPTION
**Pull request by mwildehahn**
Still testing this, but this upgrades ECS and docker and removes empire to prep running as a daemon.

### CODE REVIEW

File: ansible/roles/empire/templates/etc/empire/env/empire.env [hunk 1]
```diff
@@ -1,27 +0,0 @@
-EMPIRE_PORT=8080
```
Comment (ejholmes) [inline ansible/roles/empire/templates/etc/empire/env/empire.env:1]: I think we also need to update the [readme](https://github.com/remind101/empire_ami/blob/master/README.md)

**Comment by ejholmes** [comment]
At this point, the AMI is almost identical to the official Amazon ECS AMI (granted, that one isn't ubuntu). The only real difference here is that we include logspout.

I wonder if it would be better for us to just open source an ansible role for installing logspout. Thoughts?

File: ansible/roles/docker/templates/etc/default/docker.default [hunk 2]
```diff
@@ -4,8 +4,11 @@
 #DOCKER="/usr/local/bin/docker"
 
 # Use DOCKER_OPTS to modify the daemon startup options.
-#DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4"
-DOCKER_OPTS="--storage-driver=devicemapper"
+# We are adding the docker host IP as the dns server. This will be serviced by
+# dnsmasq running on the host. We want to do this to provide access to hosts
+# we write to /etc/hosts, in particular the dockerhost entry, where statsd will
+# be accessible from.
+DOCKER_OPTS="--dns {{ansible_default_ipv4.address}}"
```
Comment (mwildehahn) [inline ansible/roles/docker/templates/etc/default/docker.default]: this was causing:

```
docker logs -f 33fb1d9a9b94
2016-06-18T01:21:37Z [INFO] Starting Agent: Amazon ECS Agent - v1.10.0 (9a1b0f6)
2016-06-18T01:21:37Z [INFO] Loading configuration
2016-06-18T01:21:37Z [WARN] Invalid value for task cleanup duration, will be overridden to 3h0m0s, parsed value 0, minimum threshold 1m0s
2016-06-18T01:21:37Z [INFO] Checkpointing is enabled. Attempting to load state
2016-06-18T01:21:37Z [INFO] Loading state! module="statemanager"
2016-06-18T01:21:37Z [INFO] Detected Docker versions [1.17 1.18 1.19 1.20 1.21]
2016-06-18T01:21:37Z [INFO] Registering Instance with ECS
2016-06-18T01:21:58Z [ERROR] Could not register module="api client" err="RequestError: send request failed
caused by: Post https://ecs.us-west-2.amazonaws.com/: net/http: request canceled while waiting for connection"
2016-06-18T01:21:58Z [ERROR] Error registering: RequestError: send request failed
caused by: Post https://ecs.us-west-2.amazonaws.com/: net/http: request canceled while waiting for connection
```

removing this file completely fixed the issue

**Comment by phobologic** [comment]
@ejholmes So are you suggesting that we drop default logging support in the open source AMI, and make people build their own AMI with logspout, etc if they want that feature?

**Comment by ejholmes** [comment]
Pretty much. Just not sure how much value this AMI actually provides now that almost everything is stripped out.

**Comment by phobologic** [comment]
I think getting rid of logspout is kind of a big deal - it means that a really cool feature of empire won't exist for most users, and people aren't likely to want to build their own AMI. The good news is that the AMI should be much simpler after getting Empire off it, so I don't think maintaining it will be as big a deal

**Comment by phobologic** [comment]
Also, if we eventually go to the logstash based logger, we'll need to install dnsmasq as well, so we'll need some level of customization anyway.  Not sure using the ECS image is going to work well for us going forward.

**Comment by ejholmes** [comment]
Yeah that's true. Although, I was thinking we could just give some example cloudconfig to run logspout for `emp logs` support.

Logging is such a specific thing, it's hard to keep it generic enough to warrant maintaining it.

**Comment by phobologic** [comment]
Yeah, I would agree if it weren't a feature of Empire itself, and one that I think a lot of users would be bummed to lose.

The problem with using cloudconfig is that it's done on boot, so you're just asking for it to fail every once in a while and end up with hosts not behaving how you like.

**Comment by ejholmes** [comment]
Yeah true. If we stick with this, then we should aim to replace the Amazon ECS AMI in https://github.com/remind101/empire/blob/master/docs/cloudformation.json#L59-L63

**Comment by phobologic** [comment]
Yeah, fair - having them be different does sort of suck, and it'd be cool to get logging working in the demo env.

**Comment by phobologic** [comment]
@mhahn, can you take a look at this? 

https://github.com/remind101/empire_ami/pull/20/files#diff-c478b6b4b86b0ea4f627fd66bf6eef25R2

Thanks @gswallow

**Comment by mwildehahn** [comment]
@phobologic yea done

working as expected now

**Comment by mwildehahn** [comment]
actually, @gswallow, i'm still seeing the same issue when i reboot using `/var/lib`. can you verify that it's not happening for you guys even when using `/var/lib`?

**Comment by gswallow** [comment]
Which version of the agent are you using?  There's a known issue that was
fixed in ECS agent version 1.8.1.

reboot

**Comment by mwildehahn** [comment]
I'm running: 1.10.0

File: ansible/roles/docker/templates/etc/default/docker.default [hunk 3]
```diff
@@ -4,8 +4,11 @@
 #DOCKER="/usr/local/bin/docker"
 
 # Use DOCKER_OPTS to modify the daemon startup options.
-#DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4"
-DOCKER_OPTS="--storage-driver=devicemapper"
```
Comment (phobologic) [inline ansible/roles/docker/templates/etc/default/docker.default:8]: Removing this moves us back to aufs right?  I think that's the right call.

File: ansible/roles/docker/templates/etc/default/docker.default [hunk 4]
```diff
@@ -4,8 +4,11 @@
...
+# We are adding the docker host IP as the dns server. This will be serviced by
+# dnsmasq running on the host. We want to do this to provide access to hosts
+# we write to /etc/hosts, in particular the dockerhost entry, where statsd will
```
Comment (phobologic) [inline ansible/roles/docker/templates/etc/default/docker.default]: Leftover comment/opts.

**Comment by phobologic** [comment]
👍

File: ansible/roles/docker/templates/etc/default/docker.default (hunk 3 above)
Comment (mwildehahn) [inline ansible/roles/docker/templates/etc/default/docker.default:8]: Yea

```
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 26
 Dirperm1 Supported: false
```

I'm not familiar with all the different storage drivers, but after reading through the descriptions, it seems like `aufs` is the best fit for empire

**Comment by mwildehahn** [comment]
@gswallow 🙈 i hadn't swapped the ecs_agent docker image from using `/var/run`. thanks!

File: ansible/roles/docker/templates/etc/default/docker.default (hunk 3 above)
Comment (phobologic) [inline ansible/roles/docker/templates/etc/default/docker.default:8]: 👍

Comment (gswallow) [inline ansible/roles/docker/templates/etc/default/docker.default:8]: We can also make the switch to aufs.  How many docker containers do you find yourselves running on each minion, if I may ask?

**Comment by mwildehahn** [comment]
verified that reboot doesn't reregister the instance

File: ansible/roles/docker/templates/etc/default/docker.default (hunk 3 above)
Comment (phobologic) [inline ansible/roles/docker/templates/etc/default/docker.default:8]: We run around 10-15 - I think we run anywhere from 450-500 docker containers in a cluster of 35 minions (c4.2xlarge). We'd probably run more if our main API/mega-ruby-app didn't need 6 gigs of RAM to run :)

**Pull request merged**